x-pack/plugin/profiling/src/main/java/org/elasticsearch/xpack/profiling/TransportGetStackTracesAction.java

```diff
@@ -301,18 +372,56 @@ private void retrieveStackTraces(
         List<List<String>> slicedEventIds = sliced(eventIds, desiredSlices);
         ClusterState clusterState = clusterService.state();
         List<Index> indices = resolver.resolve(clusterState, "profiling-stacktraces", responseBuilder.getStart(), responseBuilder.getEnd());
+
+        // Build a set of unique host IDs.
+        Set<String> uniqueHostIDs = new HashSet<>(responseBuilder.hostEventCounts.size());
+        responseBuilder.hostEventCounts.forEach(hec -> {
+            if (hostsTable.containsKey(hec.hostID) == false) {
+                uniqueHostIDs.add(hec.hostID);
+            }
+        });
+
         StackTraceHandler handler = new StackTraceHandler(
             clusterState,
             client,
             responseBuilder,
             submitListener,
             eventIds.size(),
-            // we need to expect a set of slices for each resolved index
-            slicedEventIds.size() * indices.size()
+            // We need to expect a set of slices for each resolved index, plus one for the host metadata.
+            slicedEventIds.size() * indices.size() + (uniqueHostIDs.isEmpty() ? 0 : 1)
         );
         for (List<String> slice : slicedEventIds) {
             mget(client, indices, slice, ActionListener.wrap(handler::onResponse, submitListener::onFailure));
         }
+
+        if (uniqueHostIDs.isEmpty()) {
+            handler.calculateCO2AndCosts();
```

Comment:
I think it would be easier if we put this call into `mayFinish` and unconditionally run it on the data that are available. Then we'd instead only guard query execution here depending on whether there are new host ids. I understand that you want to parallelize execution but I expect that method to finish quickly and would rather sacrifice this for clearer code here.